storm-client/src/jvm/org/apache/storm/windowing/TimeTriggerPolicy.java

```diff
@@ -109,13 +112,22 @@ private Runnable newTriggerTask() {
         return new Runnable() {
             @Override
             public void run() {
+                // compute the current trigger timestamp based on prevTriggerTimestamp,
+                // since the calculation based on System.currentTimeMillis might have a slight drift
+                long now = System.currentTimeMillis();
+                long triggerTs = prevTriggerTimestamp == 0 ? now : prevTriggerTimestamp + duration;
+                prevTriggerTimestamp = triggerTs;
+                if (Math.abs(triggerTs - now) > 1000) {
```

Comment:
Did we think about an option to execute a bolt in a different thread to avoid timing issues in trigger because of user code?